server/remote_cache/content_addressable_storage_server/content_addressable_storage_server.go

```diff
@@ -361,7 +363,45 @@ func (s *ContentAddressableStorageServer) fetchDir(ctx context.Context, cache in
 // Errors:
 //
 // * `NOT_FOUND`: The requested tree root is not present in the CAS.
+
+type DirectoryWithDigest struct {
+	Directory *repb.Directory
+	Digest    *repb.Digest
+}
+
+func (s *ContentAddressableStorageServer) fetchDirectory(ctx context.Context, cache interfaces.Cache, dir *repb.Directory) ([]*DirectoryWithDigest, error) {
+	if len(dir.Directories) == 0 {
+		return nil, nil
+	}
+	subdirDigests := make([]*repb.Digest, 0, len(dir.Directories))
+	for _, dirNode := range dir.Directories {
+		d := dirNode.GetDigest()
+		if d.GetHash() == digest.EmptySha256 {
+			continue
+		}
+		subdirDigests = append(subdirDigests, d)
+	}
+	rspMap, err := cache.GetMulti(ctx, subdirDigests)
+	if err != nil {
+		return nil, err
+	}
+	children := make([]*DirectoryWithDigest, 0, len(subdirDigests))
+	for _, d := range subdirDigests {
+		blob, ok := rspMap[d]
+		if !ok {
+			return nil, status.NotFoundErrorf("Digest %s not found in cache.", d.GetHash())
+		}
+		subDir := &repb.Directory{}
+		if err := proto.Unmarshal(blob, subDir); err != nil {
+			return nil, err
+		}
+		children = append(children, &DirectoryWithDigest{Directory: subDir, Digest: d})
+	}
+	return children, nil
+}
+
 func (s *ContentAddressableStorageServer) GetTree(req *repb.GetTreeRequest, stream repb.ContentAddressableStorage_GetTreeServer) error {
```

Comment:
How much of a bottleneck is `GetTree`? 
IIUC, the output of GetTree is deterministic so in theory we could cache the recursed form of the tree.

Reply:
It's a significant bottleneck for certain builds that have thousands of input files and which are running on re-usable runners (where most/all of the input files already exist, but we still call GetTree). Under those conditions, we've observed GetTree calls taking upwards of 20 seconds, for basically no reason.

There are other larger fixes we can do here, but this seemed like an easy first step to improving performance for these cases since this code was basically written once (by me) and hasn't been updated since.